### PR TITLE
fix(DB/Quest): Fixed some of the Stave of the Ancients quest

### PR DESCRIPTION
**Pull request by Asthalor**
## Changes Proposed:
-  Added scripting, fixed waypoints and added gossip options to the 4 demons

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/6966

Used the scripting provided on https://github.com/TrinityCore/TrinityCore/issues/22398 with some changes to be able to update in our database

## SOURCE:

Information about the quest from here:
https://www.youtube.com/watch?v=adz-VM7wzo4

## Tests Performed:

- .go c 14527, 14529, 14531, 14536
- Check if the gossip works.
- Check if they move
- Check if they fight like they should

Small video

https://streamable.com/57dksm

## How to Test the Changes:

1. .go c 14527, 14529, 14531, 14536
2.  Check if the gossip works.
3.  Check if they move
4.  Check if they fight like they should

## Known Issues and TODO List:

- The script npc_stave_of_the_ancientsAI break this quest. Im still need to check it, but im more of a Java guys, so i might need somone to check it for me.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Review by LukasVolgger — APPROVED** [review]

**Comment by LukasVolgger** [comment]
Because this is a big change i wan't someone else to test this. Thx for this @Asthalor !

**Comment by Shybuia** [comment]
Test result:
".go creature id 14529" - Patrolling, after dialog not responding waiting to be attacked.


https://user-images.githubusercontent.com/88005118/128427206-11bf95fa-e78c-4d86-999b-d07b9aef5c50.mp4


".go creature id 14531" - Patrolling, after dialog not responding waiting to be attacked.

https://user-images.githubusercontent.com/88005118/128426528-23995998-83ed-4bd2-813b-bf224cc4be8e.mp4

".go creature id 14527" - Patrolling, after dialog not responding waiting to be attacked + pet was not patrolling with her.


https://user-images.githubusercontent.com/88005118/128426762-b20be3f2-5f06-4de1-b7be-c3b3913138bd.mp4


".go creature id 14536" - Patrolling, after dialog not responding waiting to be attacked + no loot, tried multiple times.


https://user-images.githubusercontent.com/88005118/128427333-bcb1bb26-0f9d-4dd9-94e5-ac4b206e7001.mp4

This was my first test of PR, guided by web instructions. I am confident that I did everything well from my side.

**Comment by Asthalor** [comment]
Thanks for the testing, i will take a look at those things.

**Comment by syssneck** [comment]
Hey,
I also wanted to test it later today. I will need a bit more time for this one. 
Mhh why is none of them transforming into their demon form?

**Comment by Asthalor** [comment]
> Hey,
> I also wanted to test it later today. I will need a bit more time for this one.
> Mhh why is none of them transforming into their demon form?

They take like 10 seconds to do the rp talk and transform. On that time they can be attacked, which i think they should not, so they are interrumpted i think.

**Comment by syssneck** [comment]
> 
> 
> > Hey,
> > I also wanted to test it later today. I will need a bit more time for this one.
> > Mhh why is none of them transforming into their demon form?
> 
> They take like 10 seconds to do the rp talk and transform. On that time they can be attacked, which i think they should not, so they are interrumpted i think.

Ahh, I see. Yeah, they shouldn't be attackable until they have transformed.
Okay, so then I will test it later after the changes? Ty!

**Comment by Asthalor** [comment]
Another thing ive notice is that the human forms, except Nelson can loot the heads, while on retail there are inconsistences.
For example Simone on human form should not drop it according to this page, but she does on our database,
https://classic.wowhead.com/npc=14527/simone-the-inconspicuous
but Artoriuscan drop it, according to this https://classic.wowhead.com/npc=14531/artorius-the-amiable and so does Franklin
https://classic.wowhead.com/npc=14529/franklin-the-friendly and so does Franklin

On our database only Nelson the Nice (14536) dont have the head added to his loot table, but his demon form https://classic.wowhead.com/npc=14530/solenor-the-slayer has the head.
So should we add the heads to the human forms, even tho its impossible for them to be killed without gm commands?

**Comment by Asthalor** [comment]
I fixed the errors that showed on the console, and fixed a lot of events that had a lot of errors after merging the script from Trinitycore to us. I will have to test it a bit more deeply, but i think its starting to come close to finish.

**Comment by Shybuia** [comment]
@Asthalor just let me know when it will be ready and I will test it fully, keep up the good work 👍

File: data/sql/updates/pending_db_world/rev_1627747768504565000.sql [hunk 1]
```diff
@@ -0,0 +1,748 @@
...
+UPDATE `creature` SET `MovementType` = 2 WHERE `id` IN (14536,14531,14529,14527,14503,14530,14533,14534,14535,14503);
+
+UPDATE `creature_template` SET `unit_class`= '2' WHERE `entry` IN (14533,14538);
+UPDATE `creature_template` SET `unit_class`= '1' WHERE `entry` IN (14530,14535,14534);
+
+-- The Cleaner
+UPDATE `creature_template` SET speed_run= 2.6, Mechanic_Immune_Mask= 650870623, `MovementType` = 0 WHERE entry = 14503;
+
+-- Creeping Doom
+UPDATE `creature_template` SET speed_run = 0.33, speed_walk = 0.33 WHERE entry = 14761;
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_world/rev_1627747768504565000.sql:18]: backticks

File: data/sql/updates/pending_db_world/rev_1627747768504565000.sql [hunk 2]
```diff
@@ -0,0 +1,748 @@
...
+
+(14503, 0, 0, 1, 4, 0, 100, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 'The Cleaner - On Agro Cast Fools Plight'),
+(14503, 0, 1, 0, 61, 0, 100, 0, 0, 0, 0, 0, 0, 11, 29230, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 'The Cleaner - On Agro - Cast Spell Immunity'),
+(14503, 0, 2, 0, 11, 0, 100, 0, 0, 0, 0, 0, 0, 49, 0, 0, 0, 0, 0, 0, 21, 100, 0, 0, 0, 0, 0, 0, 0, 'The Cleaner - On Spawn - Start Combat');
+
+
+UPDATE `creature_template` SET `gossip_menu_id` = '60001' WHERE `entry` = 14536;
+UPDATE `creature_template` SET `gossip_menu_id` = '5868' WHERE `entry` = 14527;
+UPDATE `creature_template` SET `gossip_menu_id` = '60003' WHERE `entry` = 14529;
+UPDATE `creature_template` SET `gossip_menu_id` = '60004' WHERE `entry` = 14531;
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_world/rev_1627747768504565000.sql:163]: Dont use single quotes around integers

**Review by Kitzunu — CHANGES_REQUESTED** [review]

File: data/sql/updates/pending_db_world/rev_1627747768504565000.sql (hunk 1 above)
Comment (Asthalor) [reply data/sql/updates/pending_db_world/rev_1627747768504565000.sql]: Added

File: data/sql/updates/pending_db_world/rev_1627747768504565000.sql (hunk 2 above)
Comment (Asthalor) [reply data/sql/updates/pending_db_world/rev_1627747768504565000.sql]: Quotes removed

**Comment by Asthalor** [comment]
Hello @Shybuia 
I think there are some things to fix here, specially the conditions that let other classes interact with the gossip (still figuring that out).
Still could you test the combats and the special things when the are struck with hunter spells and check that at least that works well? Example Simone when hit with viper sting she gets silenced.
It would help me a lot to fix the scripts if needed. Ive tested them so many times that i am probably missing something obvious.
There is no need for videos if you dont want, with written report is perfect

**Comment by Shybuia** [comment]
> Hello @Shybuia
> I think there are some things to fix here, specially the conditions that let other classes interact with the gossip (still figuring that out).
> Still could you test the combats and the special things when the are struck with hunter spells and check that at least that works well? Example Simone when hit with viper sting she gets silenced.
> It would help me a lot to fix the scripts if needed. Ive tested them so many times that i am probably missing something obvious.
> There is no need for videos if you dont want, with written report is perfect

No problem, I can try to also fix it with you.

File: data/sql/updates/pending_db_world/rev_1627747768504565000.sql [hunk 3]
```diff
@@ -0,0 +1,748 @@
...
+UPDATE `creature` SET `MovementType` = 2 WHERE `id` IN (14536,14531,14529,14527,14503,14530,14533,14534,14535,14503);
+
+UPDATE `creature_template` SET `unit_class`= '2' WHERE `entry` IN (14533,14538);
+UPDATE `creature_template` SET `unit_class`= '1' WHERE `entry` IN (14530,14535,14534);
+
+-- The Cleaner
+UPDATE `creature_template` SET `speed_run` = 2.6, `Mechanic_Immune_Mask` = 650870623, `MovementType` = 0 WHERE entry = 14503;
+
+-- Creeping Doom
+UPDATE `creature_template` SET `speed_run` = 0.33, `speed_walk` = 0.33 WHERE entry = 14761;
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_world/rev_1627747768504565000.sql:18]: backtick on entry

**Review by Kitzunu — CHANGES_REQUESTED** [review]

File: data/sql/updates/pending_db_world/rev_1627747768504565000.sql [hunk 4]
```diff
@@ -0,0 +1,748 @@
...
+-- 'npc_stave_of_the_ancients' Script is going to be removed and instead use the smart scripts
+UPDATE `creature_template` SET `AIName`='SmartAI', `MovementType` = 2 WHERE  `entry` IN (14536,14531,14529,14527,14530,14533,14534,14535);
+-- Make them patrol
+UPDATE `creature` SET `MovementType` = 2 WHERE `id` IN (14536,14531,14529,14527,14503,14530,14533,14534,14535,14503);
+
+UPDATE `creature_template` SET `unit_class`= '2' WHERE `entry` IN (14533,14538);
+UPDATE `creature_template` SET `unit_class`= '1' WHERE `entry` IN (14530,14535,14534);
+
+-- The Cleaner
+UPDATE `creature_template` SET `speed_run` = 2.6, `Mechanic_Immune_Mask` = 650870623, `MovementType` = 0 WHERE entry = 14503;
```
[the remark this answers is not in the document]
Comment (Asthalor) [reply data/sql/updates/pending_db_world/rev_1627747768504565000.sql]: Added

File: data/sql/updates/pending_db_world/rev_1627747768504565000.sql [hunk 5]
```diff
@@ -0,0 +1,748 @@
...
+-- Add the gossip option
+UPDATE `creature_template` SET `ScriptName` = '', `npcflag` = 1, `type` = 3, `faction` = 31 WHERE (`entry`IN (14527, 14529, 14531, 14536));
+UPDATE `creature_template` SET `ScriptName` = '' WHERE (`entry`IN (14538));
+-- 'npc_stave_of_the_ancients' Script is going to be removed and instead use the smart scripts
+UPDATE `creature_template` SET `AIName`='SmartAI', `MovementType` = 2 WHERE  `entry` IN (14536,14531,14529,14527,14530,14533,14534,14535);
+-- Make them patrol
+UPDATE `creature` SET `MovementType` = 2 WHERE `id` IN (14536,14531,14529,14527,14503,14530,14533,14534,14535,14503);
+
+UPDATE `creature_template` SET `unit_class`= '2' WHERE `entry` IN (14533,14538);
+UPDATE `creature_template` SET `unit_class`= '1' WHERE `entry` IN (14530,14535,14534);
```
[the remark this answers is not in the document]
Comment (Asthalor) [reply data/sql/updates/pending_db_world/rev_1627747768504565000.sql]: Removed

File: data/sql/updates/pending_db_world/rev_1627747768504565000.sql (hunk 3 above)
Comment (Asthalor) [reply data/sql/updates/pending_db_world/rev_1627747768504565000.sql]: Added

**Comment by Asthalor** [comment]
This last commit deleting the c++ script is giving a startup error:
Script named 'npc_stave_of_the_ancients' is assigned in the database, but has no code!

Using SELECT * FROM `creature_template` where `ScriptName` = 'npc_stave_of_the_ancients'; cant find any creature that use the script.

**Comment by Asthalor** [comment]
Because lately im not having that much time to fix things and because https://github.com/azerothcore/azerothcore-wotlk/pull/7929 has very good feeling im going to close this one for now.
The code is saved here in case we need to check in the future.
Thanks to the testers that helped with this.

**Pull request closed without merging**